stan/math/opencl/err/check_symmetric.hpp

```diff
@@ -31,11 +31,14 @@ inline void check_symmetric(const char* function, const char* name,
     cl::Buffer buffer_symmetric_flag(ctx, CL_MEM_READ_WRITE, sizeof(int));
     cmd_queue.enqueueWriteBuffer(buffer_symmetric_flag, CL_TRUE, 0, sizeof(int),
                                  &symmetric_flag);
-    opencl_kernels::check_symmetric(cl::NDRange(y.rows(), y.cols()), y.buffer(),
```

Comment:
@seantalts So here we actually do just pass a buffer. We could wrap this in a matrix_cl.

Is `Margs` really really really really really really  really really really  bad? Cause if it's only  really really really bad and let's us also pass in buffers then I think having two templates is fine

Reply:
I think it matters how restrictive we want the API to be. There's no real cost here of making the buffer or making a matrix_cl here. If we are worried that future us or people will start passing in buffers instead of matrix_cl types then `Margs` is bad. If we want a bit of flex then `Margs` is good

Reply:
this should work fine, actually, I think there were existing errors on the branch

Reply:
I think you or someone already changed this to take in a matrix_cl, no?

Reply:
(I believe this kernel just had too many cl::Buffers listed in the Args). But you're right, we can't just pass in buffers anymore. Are there potential uses for that?

Reply:
Hmm, if we want to do only matrix_cl's then we need to template the constructor for `matrix_cl` so it takes in type `T` since these are ints and we expect doubles everywhere else

Reply:
Though I did just refactor this and like using matrix_cl everywhere. It feels kind of cumbersome in places

Reply:
I'm confused, I didn't change `check_symmetric` to use a `matrix_cl`, it already was using that on this branch and in develop: https://github.com/stan-dev/math/blob/develop/stan/math/opencl/err/check_symmetric.hpp#L23

Reply:
Not sure if relevant, but the unpack kernel uses a simple clBuffer not part of a matrix_cl https://github.com/stan-dev/math/pull/1164/files#diff-b81a6bc2b0de6ba829ee1aa4c831c9f5R112

Reply:
hmmm yeah that is relevant. I guess Margs isn't so so so so so so bad.

I am still confused about how `check_symmetric` on develop is working even though [its kernel is declared incorrectly](https://github.com/stan-dev/math/blob/8d28f9cf1b33b16a2f96318a873a314c8a153581/stan/math/opencl/kernels/check_symmetric.hpp#L46) (two `cl::Buffer`s but there is only one passed to it).

Reply:
Sorry, I mean in the literal call to the kernel. On mobile rn, but the kernel calls variadic template sees a buffer and `to_matrix` sees that and turns it into a matix_cl in the kernel signature. So this fails to compile

Reply:
I don't think that's where the error is - I'mma wait for you to get back on desktop :P

Reply:
There are 2 buffer passed in, the second one is an integer buffer of size 1 (the flag).https://github.com/stan-dev/math/blob/8d28f9cf1b33b16a2f96318a873a314c8a153581/stan/math/opencl/err/check_symmetric.hpp

Reply:
Amiright that to_matrix in the signature of the global_kernel will turn a buffer into a matrix_cl? If so then the call to the symmetric flags kernel fails because it sees buffer_symmetric_flag, says, "this is supposed to be a matrix_cl" and then fails

Reply:
Sean what is the take on `Margs` being bad? Is is just confusing to have two variadic templates?

Reply:
I might need to add Margs back, it's true. It's bad to add the second set of template args if it's not required, because you don't need them and you end up wasting compile time and brain time to figure out what the arguments are supposed to be. You generally sort of want to constrain the inputs as much as you can with the type system. If you do need that flexibility though, it's a different story.

I think you're right about the buffer, @SteveBronder , sorry for doubting your mobile phone abilities :P

Reply:
Lol ty, but with that said I do agree have two templates is weird and not good. I have a version of this that compiles with a matrix_cl being passed to the kernel, but it throws because matrix_cl only takes in doubles and this is all integer stuff. so we need to template matrix_cls constructor and all will be well. Theres no additional cost 

I think having just matrix_cls also keeps a nice separation between our high and low level api

Reply:
Rok if you agree with the above (changing it so kernels only take in matrix_cls) can you make the unpack work with a matrix_cl?

Otherwise I bet that will be merged first and then I'll fix it up when I merge this to develop

Reply:
Done: https://github.com/stan-dev/math/pull/1164/commits/35f85c54409c90ddafa2de829e94ae39338f8b5d

Reply:
I'm happy with the new changes here, I wrote a `copy` function to pass arithmetic scalars back and forth so this code is much cleaner now!